### PR TITLE
feat: Add support of py-ios-device for certificate install on real devices

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
[py-ios-device](https://github.com/YueChen-C/py-ios-device) has several tasty features which we could integrate into Appium. For now I've just added the support for profiles installation on real devices. @KazuCocoa Could you please help with testing?

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
@YueChen-C thank you for your hard work. The library is really gorgeous.

**Comment by KazuCocoa** [comment]
np for testing. Will help.

Any reasons to pick this library? (Just for curiosity. Currently we address the below ones as part of https://github.com/appium/appium-xcuitest-driver/blob/master/README.md#optional-dependencies)

Afaik, https://github.com/alibaba/taobao-iphone-device has similar functionality to get data via the instrument.
Not the same, but https://github.com/danielpaulus/go-ios is also can print some data via com.apple.instruments. (I recently use this library, personally)

**Comment by mykola-mokhnach** [comment]
@KazuCocoa So far this tool has the most of features we'd like to add to Appium. For example performance monitoring

**Comment by YueChen-C** [comment]
Thank's for your recognition.

I know the author of **taobao-iphone-device** and have some communication with him. I open-sourced **py-ios-device** and my research direction is to get performance data on real devices. After a few months he open-sourced **taobao-iphone-device** and it has great features. For example start WDA

I plan to add the same function in the https://github.com/appium/appium-ios-device, but it was delayed by other work

**Comment by mykola-mokhnach** [comment]
> I plan to add the same function in the https://github.com/appium/appium-ios-device, but it was delayed by other work

This would be great if this functionality could be added there. 

Btw, have you ever worked with audio/video recording topic for iOS? I was wondering if it would be possible to integrate the functionality of https://github.com/danielpaulus/quicktime_video_hack into https://github.com/appium/appium-ios-device

**Comment by KazuCocoa** [comment]
>  the most of features we'd like to add to Appium. For example performance monitoring and profiles management

I see, thanks.

> https://github.com/appium/appium-xcuitest-driver/pull/1369#issuecomment-1015039665

Oh, great.

> I plan to add the same function 

👍 👍

**Comment by YueChen-C** [comment]
> Btw, have you ever worked with audio/video recording topic for iOS? I was wondering if it would be possible to integrate the functionality of https://github.com/danielpaulus/quicktime_video_hack into https://github.com/appium/appium-ios-device

Yes, I coded in Python https://github.com/YueChen-C/ios-screen-record . I think it can be integrated into Appium . But the underlying implementation is very different from https://github.com/appium/appium-ios-device

I have a problem while developing **ios-screen-record** . The usb configuration needs to be modified once before recording starts. This will disconnect the usb connection and the WDA connection. 

If the USB configuration is modified before all appium start IOS test steps. There will be no disconnection problem

File: lib/py-ios-device-client.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  async execute (args, ensureFormat = true) {
+    await this.assertExists();
+
+    const finalArgs = ['--udid', this.udid, ...args];
+    if (ensureFormat) {
+      finalArgs.push('--format', 'json');
+    }
+    log.debug(`Executing ${util.quote(this.binaryPath, finalArgs)}`);
+    const result = await exec(this.binaryPath, finalArgs);
```
Comment (KazuCocoa) [inline lib/py-ios-device-client.js:42]: need try/catch?

File: lib/commands/certificate.js [hunk 2]
```diff
@@ -309,6 +310,15 @@ commands.mobileInstallCertificate = async function mobileInstallCertificate (opt
       log.info(`The certificate cannot be installed via CLI. ` +
         `Falling back to UI-based deployment`);
     }
+  } else {
+    const client = await Pyidevice(this.opts.udid);
```
Comment (KazuCocoa) [inline lib/commands/certificate.js:314]: need `new`?

File: lib/py-ios-device-client.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    if (profilePath) {
+      await this.execute(['profiles', 'install', '--path', profilePath]);
+    } else {
+      const tmpRoot = await tempDir.openDir();
+      const tmpCertPath = path.join(tmpRoot, 'cert.pem');
+      try {
+        await fs.writeFile(tmpCertPath, payload, 'utf8');
+        await this.installProfile(tmpCertPath);
```
Comment (KazuCocoa) [inline lib/py-ios-device-client.js:67]: `{profilePath: tmpCertPath}` as the argument?

File: lib/py-ios-device-client.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  async execute (args, ensureFormat = true) {
+    await this.assertExists();
+
+    const finalArgs = ['--udid', this.udid, ...args];
+    if (ensureFormat) {
+      finalArgs.push('--format', 'json');
+    }
+    log.debug(`Executing ${util.quote(this.binaryPath, finalArgs)}`);
```
Comment (KazuCocoa) [inline lib/py-ios-device-client.js:41]: mm, this quote function seems raising `TypeError: xs.map is not a function`

File: lib/py-ios-device-client.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          `more details.`);
+      }
+      return false;
+    }
+  }
+
+  async execute (args, ensureFormat = true) {
+    await this.assertExists();
+
+    const finalArgs = ['--udid', this.udid, ...args];
```
Comment (KazuCocoa) [inline lib/py-ios-device-client.js:37]: `--udid` should be after commands such as `profiles`. So, `[...args, '--udid', this.udid];`

**Comment by KazuCocoa** [comment]
tested `mobile: installCertificate`

**Comment by KazuCocoa** [comment]
btw, i liked `notifications` command in `pyidevice`

**Comment by mykola-mokhnach** [comment]
> I have a problem while developing ios-screen-record . The usb configuration needs to be modified once before recording starts. This will disconnect the usb connection and the WDA connection.
> 
> If the USB configuration is modified before all appium start IOS test steps. There will be no disconnection problem

Yeah, this was also my concern. If we could properly setup connection mode at the very beginning when ios-device establishes USB proxy bridge then the stuff should be working as expected.

File: lib/commands/certificate.js (hunk 2 above)
Comment (mykola-mokhnach) [reply lib/commands/certificate.js:314]: fixed

File: lib/py-ios-device-client.js (hunk 3 above)
Comment (mykola-mokhnach) [reply lib/py-ios-device-client.js:67]: fixed

File: lib/py-ios-device-client.js (hunk 4 above)
Comment (mykola-mokhnach) [reply lib/py-ios-device-client.js:41]: fixed. The argument must be an array there.

File: lib/py-ios-device-client.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/py-ios-device-client.js:42]: not sure how try/catch would help here

File: lib/py-ios-device-client.js (hunk 5 above)
Comment (mykola-mokhnach) [reply lib/py-ios-device-client.js]: fixed

File: lib/py-ios-device-client.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/py-ios-device-client.js]: changed it anyway. It makes sense to provide a better error message there for a user

File: lib/py-ios-device-client.js (hunk 4 above)
Comment (KazuCocoa) [reply lib/py-ios-device-client.js]:
```
[debug] [XCUITest] Executing command 'execute'
[debug] [W3C (d7d694d2)] Encountered internal error running command: TypeError: Cannot read properties of undefined (reading 'replace')
[debug] [W3C (d7d694d2)]     at /Users/kazuaki/GitHub/appium-xcuitest-driver/node_modules/shell-quote/index.js:4:25
[debug] [W3C (d7d694d2)]     at Array.map (<anonymous>)
[debug] [W3C (d7d694d2)]     at exports.quote (/Users/kazuaki/GitHub/appium-xcuitest-driver/node_modules/shell-quote/index.js:2:15)
[debug] [W3C (d7d694d2)]     at Object.quote (/Users/kazuaki/GitHub/appium-xcuitest-driver/node_modules/appium-support/lib/util.js:340:10)
[debug] [W3C (d7d694d2)]     at Pyidevice.execute (/Users/kazuaki/GitHub/appium-xcuitest-driver/lib/py-ios-device-client.js:41:25)
[debug] [W3C (d7d694d2)]     at Pyidevice.installProfile (/Users/kazuaki/GitHub/appium-xcuitest-driver/lib/py-ios-device-client.js:72:9)
[debug] [W3C (d7d694d2)]     at XCUITestDriver.mobileInstallCertificate (/Users/kazuaki/GitHub/appium-xcuitest-driver/lib/commands/certificate.js:316:7)
[debug] [W3C (d7d694d2)]     at XCUITestDriver.executeMobile (/Users/kazuaki/GitHub/appium-xcuitest-driver/lib/commands/execute.js:158:10)
[debug] [W3C (d7d694d2)]     at XCUITestDriver.execute (/Users/kazuaki/GitHub/appium-xcuitest-driver/lib/commands/execute.js:32:12)
```

Should be `util.quote([this.binaryPath, ...finalArgs]);`?

Comment (mykola-mokhnach) [reply lib/py-ios-device-client.js]: thanks, fixed

**Review by KazuCocoa — APPROVED** [review]

**Comment by mykola-mokhnach** [comment]
Thanks for checking it @KazuCocoa 
Btw, have you tried the scenario when the same certificate is being added twice?

**Comment by KazuCocoa** [comment]
Yes. In my case, it seems like a new profile overrides the old one forcefully(?). No errors in the device UI. iOS 15.2, btw.
The result via the command was the same. (I tested via vanilla `pyidevice profiles install` as well.)

```
{
    "Status": "Acknowledged"
}
```

**Comment by kandji-joel** [comment]
@mykola-mokhnach will Appium 1.22.X receive these changes? It appears like 1.22.3 is pinned to an older version of appium-xcuitest-driver that does not include these changes

**Pull request merged**